akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -2981,16 +2981,10 @@ trait FlowOps[+Out, +Mat] {
    *
    * '''Cancels when''' downstream cancels
    */
-  def zipWithIndex: Repr[(Out, Long)] = {
-    statefulMapConcat[(Out, Long)] { () =>
-      var index: Long = 0L
-      elem => {
-        val zipped = (elem, index)
-        index += 1
-        immutable.Iterable[(Out, Long)](zipped)
-      }
-    }
-  }
+  def zipWithIndex: Repr[(Out, Long)] =
+    statefulMap(() => 0L)({
+      case (index, out) => (index + 1L, (out, index))
+    }, _ => None)
```

Comment:
Missing from the old one as well, but, should have a  name attribute.

Also, we could make it a val behind the scenes as nothing changes between separate operatior factory calls.

Reply:
Do you mean use a constantFun?

Reply:
I was thinking the whole graph, since it will be the exact same for each materialization, but maybe ignore that and leave it as is. Just add the name attribute.

Reply:
I have applied your idea